include/mc_trajectory/SequentialInterpolator.h

```diff
@@ -0,0 +1,204 @@
+   */
+  void values(const TimedValueVector & values)
+  {
+    if(values.empty()) return;
+    checkValues(values);
+    values_ = values;
+    currTime_ = values.front().first;
+    prevIndex_ = 0;
+    if(values.size() == 1)
+    {
+      nextIndex_ = 0;
+    }
+    else
+    {
+      nextIndex_ = 1;
+    }
+    intervalDuration_ = values_[nextIndex_].first - values_[prevIndex_].first;
+  }
+
+  /** Interpolation values */
+  inline const TimedValueVector & values() const noexcept
+  {
+    return values_;
+  }
+
+  /**
+   * When true, this interpolator is in a valid state and \ref compute will
+   * always return a valid value.
+   */
+  inline bool hasValues() const noexcept
+  {
+    return values_.size();
+  }
+
+  /**
+   * Clears all values
+   */
+  void clear()
+  {
+    values_.clear();
+  }
+
+  /**
+   * Compute interpolated value and move to the next element (increment time)
+   *
+   * Calls InterpolationFunction::operator(const Value &, const Value &, double ratio) where ratio is the time in the
+   * current interval normalized between 0 and 1
+   *
+   * If time is greater than the last element, return it
+   *
+   * @throws std::runtime_error If there are no values
+   */
+  Value compute()
+  {
+    if(values_.empty())
+    {
+      mc_rtc::log::error_and_throw<std::runtime_error>("SequentialInterpolator requires at least one value");
+    }
+
+    if(prevIndex_ == nextIndex_)
+    { // past last element
+      currTime_ += dt_;
+      return values_.back().second;
+    }
+    const auto & prevValue = values_[prevIndex_].second;
+    const auto & nextValue = values_[nextIndex_].second;
+    const auto & prevTime = values_[prevIndex_].first;
+    Value result = interpolator_(prevValue, nextValue, (currTime_ - prevTime) / intervalDuration_);
+    next();
+    return result;
+  }
+
+  inline double time() const noexcept
+  {
+    return currTime_;
+  }
+
+protected:
+  /** Increment time and update indices */
+  void next()
+  {
+    currTime_ += dt_;
+    const auto & nextTime = values_[nextIndex_].first;
+    if(currTime_ > nextTime)
+    {
+      if(nextIndex_ == values_.size() - 1)
+      {
+        prevIndex_ = nextIndex_;
+      }
+      else
+      {
+        prevIndex_++;
+        nextIndex_++;
+        intervalDuration_ = values_[nextIndex_].first - values_[prevIndex_].first;
+      }
+    }
+  }
+
+  /** Checks that values are strictly ordered by ascending time and spread by
+   * more than dt_ apart.
+   **/
+  void checkValues(const TimedValueVector & values)
+  {
+    double prevTime = values.front().first;
+    for(unsigned i = 1; i < values.size(); ++i)
+    {
+      auto nextTime = values[i].first;
+      if(nextTime <= prevTime || nextTime - prevTime < dt_)
+      {
+        mc_rtc::log::error_and_throw<std::runtime_error>("SequentialInterpolator values must be ordered by strictly "
+                                                         "ascending time and spread by more than dt apart");
```

Comment:
I don't think `nextTime - prevTime < dt_` has to be enforced. It's only a matter of incrementing `prevIndex_`/`nextIndex_` enough when doing the update. Down the line we could have gains specified by a tool that operates at a different frequency than the control or you could be changing the timestep on a controller but not want to update the gain ticks.

Reply:
I hadn't considered support of variable timesteps as this is something we've never done before. Regardless, I've generalized the interpolator (see message below).